### PR TITLE
Replace Action Scheduler With Shutdown Action

### PR DESCRIPTION
**Pull request by tarecord**
This removes Action Scheduler in favor of a more lightweight implementation that hooks into the "shutdown" action for sending telemetry data.

### CODE REVIEW

File: src/Telemetry/Last_Send_Subscriber.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+class Last_Send_Subscriber extends Abstract_Subscriber {
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return void
+	 */
+    public function register() {
```
Comment (borkweb) [inline src/Telemetry/Last_Send_Subscriber.php:29]:
```suggestion
	public function register() {
```

File: src/Telemetry/Routes/Check.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+class Check extends Abstract_Route {
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return void
+	 */
+    public function get_endpoint() {
```
Comment (borkweb) [inline src/Telemetry/Routes/Check.php:30]:
```suggestion
	public function get_endpoint() {
```

File: src/Telemetry/Routes/Check.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param WP_REST_Request $request
+	 *
+	 * @return void
+	 */
+    public function action( WP_REST_Request $request ) {
```
Comment (borkweb) [inline src/Telemetry/Routes/Check.php:43]:
```suggestion
	public function action( WP_REST_Request $request ) {
```

File: src/Telemetry/Routes/Route_Subscriber.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+class Route_Subscriber extends Abstract_Subscriber {
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return void
+	 */
+    public function register() {
```
Comment (borkweb) [inline src/Telemetry/Routes/Route_Subscriber.php:30]:
```suggestion
	public function register() {
```

File: src/Telemetry/Routes/Send.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+class Send extends Abstract_Route {
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @return void
+	 */
+    public function get_endpoint() {
```
Comment (borkweb) [inline src/Telemetry/Routes/Send.php:32]:
```suggestion
	public function get_endpoint() {
```

File: src/Telemetry/Routes/Send.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	/**
+	 * @inheritDoc
+	 *
+	 * @since 1.0.0
+	 *
+	 * @param WP_REST_Request $request
+	 *
+	 * @return void
+	 */
+    public function action( WP_REST_Request $request ) {
```
Comment (borkweb) [inline src/Telemetry/Routes/Send.php:45]:
```suggestion
	public function action( WP_REST_Request $request ) {
```

File: src/Telemetry/Telemetry_Subscriber.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+ * @since 1.0.0
+ */
+class Telemetry_Subscriber extends Abstract_Subscriber {
+
+	/**
+	 * @inheritDoc
+	 *
+	 * @return void
+	 */
+    public function register() {
```
Comment (borkweb) [inline src/Telemetry/Telemetry_Subscriber.php:23]:
```suggestion
	public function register() {
```

File: src/Telemetry/Telemetry_Subscriber.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+			return;
+		}
+
+		$nonce           = wp_create_nonce( Telemetry::NONCE );
+		$route_namespace = $this->container->get( Send::class )->get_namespace();
+
+		try {
+			wp_remote_get( get_rest_url( null, $route_namespace . '/send?nonce=' . $nonce . '&timestamp=' . $timestamp ), [
+				'blocking' => false,
+				'timeout' => 1
```
Comment (borkweb) [inline src/Telemetry/Telemetry_Subscriber.php:57]:
```suggestion
				'timeout' => 1,
```

File: src/Telemetry/Routes/Send.php [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+			return;
+		}
+
+		if ( ! wp_verify_nonce( $request->get_param( 'nonce' ), Telemetry::NONCE ) ) {
+			$this->send_early_unauthorized();
+		}
+
+		Config::get_container()->get( Telemetry::class )->send_data();
+
+		$this->send_early_ok();
```
Comment (borkweb) [inline src/Telemetry/Routes/Send.php:61]: We want to flip these two, that way we respond with an early OK once we've validated the nonce, and then do the work to send telemetry data.

```suggestion
		$this->send_early_ok();

		Config::get_container()->get( Telemetry::class )->send_data();
```

**Review by borkweb — CHANGES_REQUESTED** [review]

File: src/Telemetry/Core.php [hunk 10]
```diff
@@ -30,9 +30,10 @@ class Core {
 	 * @var string[]
 	 */
 	private array $subscribers = [
-		Cron_Subscriber::class,
-		Opt_In_Subscriber::class,
 		Exit_Interview_Subscriber::class,
+		Last_Send_Subscriber::class,
+		Opt_In_Subscriber::class,
+		Telemetry_Subscriber::class,
 		Admin_Subscriber::class,
```
Comment (ChrisMKindred) [inline src/Telemetry/Core.php:37]: 🟠 
We should keep arrays like this in alphabetical order to make the classes easier to find in the list.

File: src/Telemetry/Telemetry.php [hunk 11]
```diff
@@ -19,6 +19,7 @@
  */
 class Telemetry {
 	public const SERVER_URL = 'https://telemetry-api.moderntribe.qa/api/v1';
+	public const AJAX_ACTION = 'stellarwp_telemetry_send_data';
```
Comment (ChrisMKindred) [inline src/Telemetry/Telemetry.php:22]: 🔴 
Align the equal signs
```suggestion
	public const SERVER_URL  = 'https://telemetry-api.moderntribe.qa/api/v1';
	public const AJAX_ACTION = 'stellarwp_telemetry_send_data';
```

**Review by ChrisMKindred — CHANGES_REQUESTED** [review]
A couple of small format items. Functionally looks great.

**Pull request merged**